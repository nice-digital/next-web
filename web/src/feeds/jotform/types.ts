export type FormID = `${number}`;

export type Duration = `${number}ms`;

export type NotFoundResponse = {
	responseCode: 404;
	message: string;
	content: "";
	duration: Duration;
	info: string;
};

export type NotAuthorizedResponse = {
	responseCode: 401;
	message: string;
	content: "";
	duration: Duration;
	info: string;
};

export type FormStatus = "ENABLED" | "DISABLED";

export type FormType = "LEGACY" | "CARD";

export type FormSuccessResponse = {
	responseCode: 200;
	message: "success";
	content: {
		id: string;
		username: string;
		title: string;
		height: number;
		status: FormStatus;
		/** Format like "2023-03-21 09:27:18" */
		created_at: string;
		updated_at: string;
		last_submission: string;
		new: `${number}`;
		count: `${number}`;
		type: FormType;
		favorite: `${number}`;
		archived: `${number}`;
		url: string;
	};
	duration: Duration;
};

export type GetFormResponse =
	| NotAuthorizedResponse
	| NotFoundResponse
	| FormSuccessResponse;
