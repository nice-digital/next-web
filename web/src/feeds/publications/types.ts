import type { Except, ReadonlyDeep } from "type-fest";

export enum FeedPath {
	ProductsLite = "/feeds/products-lite",
	ProductTypes = "/feeds/producttypes",
	AreasOfInterest = "/feeds/areaofinteresttypes",
	ProductDetail = "/feeds/product/",
}

export enum ProductStatus {
	Published = "Published",
	Withdrawn = "Withdrawn",
}

export enum ProductGroup {
	Guideline = "Guideline",
	Guidance = "Guidance",
	Advice = "Advice",
	Standard = "Standard",
	Corporate = "Corporate",
	Other = "Other",
}

export const enum ProductTypeAcronym {
	/** Antimicrobial prescribing evidence summaries */
	"APA" = "APA",
	/** Antimicrobial prescribing guidelines */
	"APG" = "APG",
	/** Clinical guidelines */
	"CG" = "CG",
	/** COVID-19 rapid evidence summaries */
	"COA" = "COA",
	/** COVID-19 rapid guidelines */
	"COV" = "COV",
	/** COVID-19 rapid MIBs */
	"COVM" = "COVM",
	/** COVID-19 tests */
	"COVT" = "COVT",
	/** Cancer service guidelines */
	"CSG" = "CSG",
	/** Diagnostics guidance */
	"DG" = "DG",
	/** Corporate documents (corporate)  */
	"ECD" = "ECD",
	/** Evidence summaries (advice) */
	"ES" = "ES",
	/** Highly specialised technologies guidance */
	"HST" = "HST",
	/** Indicators */
	"IND" = "IND",
	/** Interventional procedures guidance */
	"IPG" = "IPG",
	/** Key therapeutic topics (advice) */
	"KTT" = "KTT",
	/** Local government briefing (advice) */
	"LGB" = "LGB",
	/** Medtech innovation briefings (advice) */
	"MIB" = "MIB",
	/** Medicines practice guidelines */
	"MPG" = "MPG",
	/** Medical technologies guidance */
	"MTG" = "MTG",
	/** NICE guidelines */
	"NG" = "NG",
	/** Public health guidelines */
	"PH" = "PH",
	/** Public involvement programmes (advice) */
	"PIP" = "PIP",
	/** Process and methods guides (corporate) */
	"PMG" = "PMG",
	/** Patient safety guidance */
	"PSG" = "PSG",
	/** Quality standards (standards) */
	"QS" = "QS",
	/** Social care guidelines */
	"SC" = "SC",
	/** Safe staffing guidelines */
	"SG" = "SG",
	/** Technology appraisal guidance */
	"TA" = "TA",
}

export enum AreaOfInterestAcronym {
	/**
	 * Covid-19
	 */
	COV = "COV",
	/**
	 * Antimicrobial prescribing
	 */
	AMP = "AMP",
}

/** Some product types can act as parents for sub-types */
export type ParentProductTypeAcronym =
	// ES (Evidence summaries) have sub types like APA (Antimicrobial prescribing evidence summaries) and COA (COVID-19 rapid evidence summaries)
	| ProductTypeAcronym.ES
	// MIBs (Medtech innovation briefings) have sub types like COA (COVID-19 rapid MIBs) and COVT (COVID-19 tests)
	| ProductTypeAcronym.MIB;

/** Only certain types can be used for 'developed as' */
export type DevelopedAsProductTypeAcronym =
	| ProductTypeAcronym.APA
	| ProductTypeAcronym.APG
	| ProductTypeAcronym.CG
	| ProductTypeAcronym.COA
	| ProductTypeAcronym.COV
	| ProductTypeAcronym.COVM
	| ProductTypeAcronym.MPG
	| ProductTypeAcronym.PH
	| ProductTypeAcronym.SC
	| ProductTypeAcronym.SG;

/** Only certain types can be used for 'relevant to' */
export type RelevantToProductTypeAcronym =
	| ProductTypeAcronym.APA
	| ProductTypeAcronym.APG
	| ProductTypeAcronym.COV
	| ProductTypeAcronym.CSG
	| ProductTypeAcronym.MPG
	| ProductTypeAcronym.PH
	| ProductTypeAcronym.SC
	| ProductTypeAcronym.SG;

export type Link = {
	readonly href: string;
};

interface EmptyLinks {
	/** Empty object that's never used but in here for completeness */
	readonly self: [Record<string, never>];
}

export type BaseFeedItem = ReadonlyDeep<{
	_links: EmptyLinks;
	/** ETag is always null so kind of pointless but kept here for completeness */
	eTag: null;
	/** Full ISO date string like `2021-04-15T08:18:13.7945978Z` */
	lastModified: string;
}>;

/**
 * The raw object that comes back from the feed
 */
export type ProductLiteRaw = BaseFeedItem &
	ReadonlyDeep<{
		_links: EmptyLinks & {
			"nice.publications:productfeed": Link[];
		};
		ETag: null;
		Id: string;
		Title: string;
		ProductStatus: ProductStatus;
		ProductType: ProductTypeAcronym;
		/** ISO date string like `2017-07-06T00:00:00`. Notice it's a rounded time because it's set manually by an editor. */
		PublishedDate: string;
		/** ISO date string like `2017-07-06T00:00:00` or `2021-05-21T10:54:52.4655487`. */
		LastMajorModificationDate: string;
		AreasOfInterestList: AreaOfInterestAcronym[];
		DevelopedAs: DevelopedAsProductTypeAcronym | null;
		RelevantTo: RelevantToProductTypeAcronym[];
		ProductGroup: ProductGroup;
	}>;

/** A product lite from the feed, but with redundant properties removed */
export type ProductLite = Except<ProductLiteRaw, "ETag" | "_links">;

export type ProductType = BaseFeedItem &
	ReadonlyDeep<{
		enabled: boolean;
		name: string;
		pluralName: string;
		identifierPrefix: ProductTypeAcronym;
		group: ProductGroup;
		parent: ParentProductTypeAcronym | "";
	}>;

export type AreaOfInterest = BaseFeedItem &
	ReadonlyDeep<{
		Enabled: boolean;
		Name: string;
		PluralName: string;
	}>;

type ETag = string | null;

// axios-case-converter maps key property names like nice.publications:area-of-interest-type-list to nicePublicationsAreaOfInterestTypeList
type EmbeddedKey = `nicePublications${string}`;

type Embedded<TKey extends EmbeddedKey, TInner> = {
	embedded: { [key in TKey]: TInner };
};

type FeedContentInner<TEmbeddedInner extends EmbeddedKey, TItemType> = {
	_links: {
		self: [Link];
	};
	eTag: ETag;
} & Embedded<TEmbeddedInner, TItemType[]>;

type FeedContent<
	TEmbeddedOuter extends EmbeddedKey,
	TEmbeddedInner extends EmbeddedKey,
	TItemType
> = {
	_links: {
		self: [Link];
	};
	eTag: ETag;
	lastModified: string;
} & Embedded<TEmbeddedOuter, FeedContentInner<TEmbeddedInner, TItemType>>;

export type AreasOfInterestList = FeedContent<
	"nicePublicationsAreaOfInterestTypeList",
	"nicePublicationsAreaOfInterestType",
	AreaOfInterest
>;

export type ProductTypeList = FeedContent<
	"nicePublicationsProductTypeList",
	"nicePublicationsProductType",
	ProductType
>;

export type ProductListLite = FeedContent<
	"nicePublicationsProductListLite",
	"nicePublicationsProductLite",
	ProductLiteRaw
>;

export type ProductChapter = {
	title: string;
	url: string;
};

export type HTMLChapterContentInfo = {
	title: string;
	chapterSlug: string;
	_links: { self: [Link] };
};

export type HTMLChapterContent = {
	eTag: string;
	content: string;
	title: string;
	reference: string;
	partId: number;
	chapterSlug: string;
};

export type HTMLContent = Embedded<
	"nicePublicationsHtmlChapterContentInfo",
	HTMLChapterContentInfo[]
>;

// export type HTMLContent = {
// 	_embedded: {
// 		"nice.publications:html-chapter-content-info": HTMLChapterContentInfo[];
// 	};
// };

export type FileContent<TFileExtension extends "pdf" | "mobi" | "epub"> = {
	fileName: `${string}.${TFileExtension}`;
	length: number;
	_links: { self: [Link] };
	eTag: string | null;
	uid: number;
	id: string;
	mimeType: string;
	hash: string;
	name: string;
};

export type UploadAndConvertContentPart = {
	embedded: {
		nicePublicationsHtmlContent: HTMLContent;
		nicePublicationsPdfFile: FileContent<"pdf">;
		nicePublicationsMobiFile: FileContent<"mobi">;
		nicePublicationsEpubFile: FileContent<"epub">;
	};
};

// export type ContentPartListOLD = {
// 	"nice.publications:content-part-list": {
// 		embedded: UploadAndConvertContentPart;
// 	};
// };

export type ContentPartList = Embedded<
	"nicePublicationsUploadAndConvertContentPart",
	UploadAndConvertContentPart
>;

export type ProductDetail = {
	chapterHeadings?: ProductChapter[];
	// embedded: ContentPartList;
	id: string;
	lastUpdatedDate?: string;
	productType: string;
	publishedDate?: string;
	summary?: string;
	title: string;
} & Embedded<"nicePublicationsContentPartList", ContentPartList>;

export type ErrorResponse = {
	StatusCode: string;
	Message: string;
};
