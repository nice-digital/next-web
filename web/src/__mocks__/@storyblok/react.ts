export const storyblokInit = (): null => null;

export const apiPlugin = (): null => null;
