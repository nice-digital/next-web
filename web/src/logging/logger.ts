// import { useRouter } from "next/router";
import { useRouter } from "next/compat/router";
import pino, { Logger } from "pino";

import { niceLoggingPinoOptions } from "./nice-logging";

export const logger = pino(niceLoggingPinoOptions);

/**
 * React hook for using a logger within a React component (or other hook) that logs extra info to do with the current route.
 *
 * @returns A child pino logger that automatically logs extra routing info
 */
export const useLogger = (): Logger => {
	const router = useRouter();

	const pathname = router?.pathname;
	const query = router?.query;
	// 'Enrich' logs with extra request info like we do here in Guidance Web:
	// https://github.com/nice-digital/guidance-web/blob/master/Guidance.Web/RequestEnricher.cs#L19-L26
	return logger.child({
		Properties: {
			RequestPath: pathname,
			RequestQueryString: JSON.stringify(query), // Types have to match in ELK so make sure this is a string and not an object
			RequestUserAgent:
				typeof window === "undefined" ? undefined : window.navigator.userAgent,
		},
	});
};
