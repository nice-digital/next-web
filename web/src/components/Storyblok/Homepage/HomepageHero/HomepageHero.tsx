import { useState, useEffect } from "react";

import { Button } from "@nice-digital/nds-button";
import { Hero } from "@nice-digital/nds-hero";

import { type HomepageHeroStoryblok } from "@/types/storyblok";

interface HomepageHeroBlokProps {
	blok: HomepageHeroStoryblok;
}

export const HomepageHero = ({
	blok,
}: HomepageHeroBlokProps): React.ReactNode => {
	const { title, images, description, ctaLink, ctaText } = blok;

	const [randomImage, setRandomImage] = useState("");

	// Pick an image at random
	useEffect(() => {
		const randomIndex = Math.floor(Math.random() * images.length);
		setRandomImage(`${images[randomIndex].filename}/m/`);
	}, [images]);

	return (
		<Hero
			isDark
			title={title}
			intro={description}
			image={randomImage}
			actions={
				<Button variant="cta" to={`/${ctaLink.cached_url}`}>
					{ctaText}
				</Button>
			}
		/>
	);
};
