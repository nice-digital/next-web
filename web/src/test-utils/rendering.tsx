/* eslint-disable import/export, import/no-unresolved -- We're deliberately overrind the default render from @testing-library with a custom render method */
import {
	render as testingLibrRender,
	RenderResult,
} from "@testing-library/react";
import { HeadManagerContext } from "next/dist/shared/lib/head-manager-context.shared-runtime";
import { RouterContext } from "next/dist/shared/lib/router-context.shared-runtime";
import { NextRouter } from "next/router";
import { ReactElement, useEffect } from "react";
import { renderToString } from "react-dom/server";

import NextWebApp from "../pages/_app";

import type { AppProps } from "next/app";

/**
 *
 */
export const mockRouter: NextRouter = {
	basePath: "",
	pathname: "/",
	route: "/",
	asPath: "/",
	query: {},
	push: jest.fn(),
	replace: jest.fn(),
	reload: jest.fn(),
	back: jest.fn(),
	forward: jest.fn(),
	prefetch: jest.fn(),
	beforePopState: jest.fn(),
	events: {
		on: jest.fn(),
		off: jest.fn(),
		emit: jest.fn(),
	},
	isFallback: false,
	isLocaleDomain: false,
	isPreview: false,
	isReady: false,
};

/**
 * Convenience method for getting a mock router with the given properties.
 *
 * @example
 * 	getMockRouter({ pathname: "/about" })
 *
 * @param routerParams Optional router properties that get merged with the default mock router
 * @returns A new mock router object
 */
export const getMockRouter = (
	routerParams?: Partial<NextRouter>
): NextRouter => ({
	...mockRouter,
	...routerParams,
});

/**
 * Collect tags from next/head: https://edibleco.de/3Anu92S
 */
const HeadProvider: React.FC = ({ children }) => {
	let head: JSX.Element[];

	useEffect(() => {
		global.document.head.insertAdjacentHTML(
			"afterbegin",
			renderToString(<>{head}</>) || ""
		);
	});

	return (
		<HeadManagerContext.Provider
			value={{
				updateHead: (state) => {
					head = state;
				},
				mountedInstances: new Set(),
			}}
		>
			{children}
		</HeadManagerContext.Provider>
	);
};

export type GetWrapperOptions = {
	router: NextRouter;
	wrapper: RenderOptions["wrapper"];
};
const getAppWrapper =
	({ router, wrapper: Wrapper }: GetWrapperOptions) =>
	({ children }: { children: ReactElement }) =>
		(
			<HeadProvider>
				<RouterContext.Provider value={router}>
					<NextWebApp
						Component={() => (
							<>{Wrapper ? <Wrapper>{children}</Wrapper> : children}</>
						)}
						pageProps={{}}
						router={router as AppProps["router"]}
					/>
				</RouterContext.Provider>
			</HeadProvider>
		);

export type RenderOptions = Parameters<typeof testingLibrRender>[1] & {
	router?: Partial<NextRouter>;
};

/**
 * Custom render method the given component, wrapped in `NextWebApp` and a `HeadProvider`.
 *
 * @see https://testing-library.com/docs/react-testing-library/setup/#custom-render
 *
 * @param ui The element to render
 * @param options Optional render options that get passed to underlying React Testing Lib render method
 * @returns The render result
 */
const customRender = (
	ui: ReactElement,
	{ wrapper, router, ...options }: RenderOptions = {}
): RenderResult =>
	testingLibrRender(ui, {
		wrapper: getAppWrapper({ router: getMockRouter(router), wrapper }),
		...options,
	});

export * from "@testing-library/react";
export { customRender as render };
