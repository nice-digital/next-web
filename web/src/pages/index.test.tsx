import { ISbStoryData } from "@storyblok/react";
import { GetServerSidePropsContext } from "next";

import { homepageHeroProps } from "@/components/Storyblok/Homepage/HomepageHero/HomepageHero.test";
import { logger } from "@/logger";
import { render, screen } from "@/test-utils/rendering";
import {
	mockStoryblokStory,
	mockNewsArticle,
	mockBlogPost,
	mockRichText,
} from "@/test-utils/storyblok-data";
import mockLatestNews from "@/test-utils/storyblok-homepage-latestnews-response.json";
import mockResult from "@/test-utils/storyblok-homepage-result.json";
import mockFetchStory from "@/test-utils/storyblok-homepage-storyResult-response.json";
import MockServerErrorResponse from "@/test-utils/storyblok-server-error-response.json";
import { NewsStory } from "@/types/News";
import * as storyblokUtils from "@/utils/storyblok";
import { GENERIC_ERROR_MESSAGE } from "@/utils/storyblok";

import Home, { type HomePageProps, getServerSideProps } from "./index.page";

// Mock a second article with a different ID so we avoid duplicate key warnings
const secondNewsArticle = {
	...mockNewsArticle,
	id: 987654321,
};

const latestNews: ISbStoryData<NewsStory>[] = [
	mockNewsArticle,
	mockBlogPost,
	secondNewsArticle,
];

const props: HomePageProps = {
	latestNews,
	story: {
		...mockStoryblokStory,
		content: {
			hero: [homepageHeroProps.blok],
			links: mockRichText,
			_uid: "12345789",
			component: "homepage",
		},
	},
};

const expectedErrorMessage = GENERIC_ERROR_MESSAGE;

describe("Homepage", () => {
	it("should match snapshot for main content", () => {
		const { container } = render(<Home {...props} />);
		expect(container).toMatchSnapshot();
	});

	it("Should render error content when error is passed as a prop", async () => {
		const errorMessage = "An error returned from getserversideprops";
		const { asFragment } = render(<Home error={errorMessage} />);

		expect(screen.getByText(errorMessage)).toBeInTheDocument();
		expect(asFragment()).toMatchSnapshot();
	});

	describe("getServerSideProps", () => {
		let fetchStoriesSpy: jest.SpyInstance;
		let fetchStorySpy: jest.SpyInstance;
		beforeEach(() => {
			fetchStoriesSpy = jest.spyOn(storyblokUtils, "fetchStories");
			fetchStorySpy = jest.spyOn(storyblokUtils, "fetchStory");
			jest.useFakeTimers();
			jest.setSystemTime(new Date("2024-04-08"));
		});

		afterEach(() => {
			jest.clearAllMocks();
		});

		it("should return error when fetchStory returns error", async () => {
			const mockErrorMessage =
				"<<<< There was an error fetching stories. Please try again later. >>>>> ";
			const mockCatchError = new Error(mockErrorMessage);

			fetchStorySpy.mockRejectedValueOnce(mockCatchError);

			const context = {
				query: {},
				params: { slug: "home" },
				req: { headers: {} },
			} as unknown as GetServerSidePropsContext;

			const result = await getServerSideProps(context);

			expect(result).toEqual({ props: { error: expectedErrorMessage } });
		});

		it("should call fetchStories and fetchStory", async () => {
			fetchStoriesSpy.mockResolvedValueOnce(mockLatestNews);
			fetchStorySpy.mockResolvedValueOnce(mockFetchStory);

			const context = {
				query: {},
				params: { slug: "home" },
				req: { headers: {} },
			} as unknown as GetServerSidePropsContext;

			const result = await getServerSideProps(context);

			expect(fetchStorySpy).toHaveBeenCalledTimes(1);
			expect(fetchStoriesSpy).toHaveBeenCalledTimes(1);
			expect(fetchStoriesSpy).toHaveBeenCalledWith("published", {
				excluding_ids: "430633954",
				excluding_slugs: "news/blogs/authors/*",
				filter_query: {
					date: {
						lt_date: "2024-04-08T00:00:00.000Z",
					},
				},
				per_page: 3,
				resolve_relations: "blogPost.author",
				sort_by: "content.date:desc",
				starts_with: "news",
			});

			expect(result).toEqual(mockResult);
		});

		it("should log and error if validateRouteParams throws an error", async () => {
			expect(jest.isMockFunction(logger.error)).toBe(true);
			const mockErrorMessage =
				"There was an error fetching this story. Please try again later.";
			const mockCatchError = new Error(mockErrorMessage, {
				cause: MockServerErrorResponse,
			});

			fetchStorySpy.mockRejectedValueOnce(mockCatchError);

			const mockContext = {
				query: {},
				params: { slug: "home" },
				req: {
					headers: {
						"cache-control":
							"public, s-max-age=300, max-age=120, stale-while-revalidate=1800",
					},
				},
			} as unknown as GetServerSidePropsContext;

			await getServerSideProps(mockContext);

			expect(logger.error).toHaveBeenCalled();
			expect(logger.error).toHaveBeenCalledWith(
				{
					errorCause: mockCatchError.cause,
					requestHeaders: mockContext.req.headers,
				},
				`Error fetching Homepage from gssp`
			);
		});

		it("should return error when fetchStories returns error", async () => {
			const mockErrorMessage =
				"<<<< There was an error fetching stories. Please try again later. >>>>> ";
			const mockCatchError = new Error(mockErrorMessage);

			fetchStoriesSpy.mockRejectedValueOnce(mockCatchError);

			const context = {
				query: {},
				params: { slug: "home" },
				req: { headers: {} },
			} as unknown as GetServerSidePropsContext;

			const result = await getServerSideProps(context);

			expect(result).toEqual({ props: { error: expectedErrorMessage } });
		});
	});
});
