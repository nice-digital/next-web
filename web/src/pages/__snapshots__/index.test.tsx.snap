// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Homepage Hero should match snapshot 1`] = `
<div>
  <div
    class="undefined undefined"
  >
    <div />
    <main>
      <div
        class="container"
        data-component="container"
      >
        <div
          class="hero hero--dark hero"
          data-component="hero"
        >
          <div
            class="hero__container"
          >
            <div
              class="hero__body"
            >
              <h1
                class="hero__title"
              >
                Holding out for a homepage hero
              </h1>
              <p
                class="hero__intro"
              >
                'til the end of the night
              </p>
              <div
                class="hero__actions"
              >
                <a
                  class="btn btn--cta"
                  data-component="button--cta"
                  href="https://www.youtube.com/watch?v=bWcASV2sey0&pp=ygUjYm9ubmllIHR5bGVyIGhvbGRpbmcgb3V0IGZvciBhIGhlcm8%3D"
                >
                  Click this thing
                </a>
              </div>
            </div>
          </div>
          <div
            class="hero__image-container"
            style="background-image: url(http://local-host-test-nice-org.com/f/243782/420x234/7a576133e7/register-for-board-meeting.PNG/m/867x0/filters:format%28jpeg%29:quality%2880%29/m/);"
          />
        </div>
      </div>
    </main>
    <noscript />
  </div>
</div>
`;

exports[`Homepage Should render error content when error is passed as a prop 1`] = `
<DocumentFragment>
  <div
    class="undefined undefined"
  >
    <div />
    <main>
      <div
        class="container"
        data-component="container"
      >
        <nav
          aria-label="Breadcrumbs"
          data-component="breadcrumbs"
          role="navigation"
        >
          <ol
            class="breadcrumbs"
          >
            <li
              class="breadcrumbs__crumb"
            >
              <a
                href="/"
              >
                Home
              </a>
            </li>
            <li
              class="breadcrumbs__crumb"
            >
              <span>
                Error
              </span>
            </li>
          </ol>
        </nav>
        <script
          type="application/ld+json"
        >
          {
  "@context": "http://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position": 1,
      "item": {
        "@id": "/",
        "name": "Home"
      }
    },
    {
      "@type": "ListItem",
      "position": 2,
      "item": {
        "name": "Error"
      }
    }
  ]
}
        </script>
        <div
          class="page-header  page-header--vertical-padding-loose"
          data-component="page-header"
          id="content-start"
        >
          <h1
            class="page-header__heading "
          >
            An error returned from getserversideprops
          </h1>
          <p
            class="page-header__lead"
          >
            We'll look into it right away. Please try again in a few minutes. And if it's still not fixed, 
            <a
              href="/get-involved/contact-us"
            >
              contact us
            </a>
            .
          </p>
        </div>
        <h2>
          What can I doÂ now?
        </h2>
        <div
          class="grid grid--loose"
        >
          <div
            data-component="grid-item"
            data-g="12 sm:6 md:4"
          >
            <ul
              class="mt--0"
            >
              <li>
                Check that the web address has been typed correctly
              </li>
              <li>
                Look for it using search
              </li>
              <li>
                Browse for it from the 
                <a
                  href="/"
                >
                  homepage
                </a>
              </li>
            </ul>
            <p>
              And if you still can't find what you're looking for, 
              <a
                href="/get-involved/contact-us"
              >
                get in touch
              </a>
              .
            </p>
          </div>
          <div
            data-component="grid-item"
            data-g="12 sm:6 md:8"
          >
            <p
              class="mt--0"
            >
              Here's where you can go from here:
            </p>
            <ul>
              <li>
                Explore 
                <a
                  href="/guidance"
                >
                  NICE guidance, quality standards and advice
                </a>
                . You can do this by 
                <a
                  href="/guidance/conditions-and-diseases"
                >
                  category to find a topic
                </a>
                , or choose a 
                <a
                  href="/guidance/published"
                >
                  list organised by date or title
                </a>
                .
              </li>
              <li>
                <a
                  href="/about"
                >
                  Visit our about pages
                </a>
                 to find information about our services, 
                <a
                  href="/about/who-we-are"
                >
                  who we are
                </a>
                 and 
                <a
                  href="/about/what-we-do"
                >
                  what we do
                </a>
              </li>
            </ul>
            <p>
              If this is not the first time you have seen this message and believe that there is a serious problem 
              <a
                href="/get-involved/contact-us"
              >
                please contact us directly
              </a>
            </p>
          </div>
        </div>
      </div>
    </main>
    <noscript />
  </div>
</DocumentFragment>
`;

exports[`Homepage should match snapshot for main content 1`] = `
<div>
  <div
    class="undefined undefined"
  >
    <div />
    <main>
      <div
        class="container"
        data-component="container"
      >
        <div
          class="hero hero--dark hero"
          data-component="hero"
        >
          <div
            class="hero__container"
          >
            <div
              class="hero__body"
            >
              <h1
                class="hero__title"
              >
                Holding out for a homepage hero
              </h1>
              <p
                class="hero__intro"
              >
                'til the end of the night
              </p>
              <div
                class="hero__actions"
              >
                <a
                  class="btn btn--cta"
                  data-component="button--cta"
                  href="https://www.youtube.com/watch?v=bWcASV2sey0&pp=ygUjYm9ubmllIHR5bGVyIGhvbGRpbmcgb3V0IGZvciBhIGhlcm8%3D"
                >
                  Click this thing
                </a>
              </div>
            </div>
          </div>
          <div
            class="hero__image-container"
            style="background-image: url(http://local-host-test-nice-org.com/f/243782/420x234/7a576133e7/register-for-board-meeting.PNG/m/867x0/filters:format%28jpeg%29:quality%2880%29/m/);"
          />
        </div>
        <section
          class="latestNews"
          data-tracking="latest-news"
        >
          <h2
            class="title"
          >
            Latest from NICE
          </h2>
          <ol
            class="newsGrid"
          >
            <li
              class="newsItem"
            >
              <article
                class="newsCard "
              >
                <a
                  aria-hidden="true"
                  class="imageContainer"
                  href="/i/am/a-very-full-slug"
                  style="background-image: url(http://local-host-test-nice-org.com/f/243782/420x234/7a576133e7/register-for-board-meeting.PNG/m/867x0/filters:format%28jpeg%29:quality%2880%29/m/868x0/filters:quality%2880%29);"
                  tabindex="-1"
                >
                   
                </a>
                <div>
                  <h3
                    class="heading"
                  >
                    <a
                      href="/i/am/a-very-full-slug"
                    >
                      Test news article title
                    </a>
                  </h3>
                  <p>
                    Test intro text
                  </p>
                  <footer>
                    <div>
                      <span
                        class="tag tag--outline"
                        data-component="tag"
                      >
                        News
                      </span>
                      <time
                        class="date"
                        datetime="2024-01-31T15:16:15.19"
                      >
                        31 January 2024
                      </time>
                    </div>
                  </footer>
                </div>
              </article>
            </li>
            <li
              class="newsItem"
            >
              <article
                class="newsCard "
              >
                <a
                  aria-hidden="true"
                  class="imageContainer"
                  href="/i/am/a-very-full-slug"
                  style="background-image: url(http://local-host-test-nice-org.com/f/243782/420x234/7a576133e7/register-for-board-meeting.PNG/m/867x0/filters:format%28jpeg%29:quality%2880%29/m/868x0/filters:quality%2880%29);"
                  tabindex="-1"
                >
                   
                </a>
                <div>
                  <h3
                    class="heading"
                  >
                    <a
                      href="/i/am/a-very-full-slug"
                    >
                      Test blog post title
                    </a>
                  </h3>
                  <p
                    class="blogIntroText"
                  >
                    Let's play Blockbusters!
                  </p>
                  <footer>
                    <div
                      class="author"
                    >
                      Bob Holness
                      , 
                      Blockbusters host
                    </div>
                    <div>
                      <span
                        class="tag tag--outline"
                        data-component="tag"
                      >
                        Blogs
                      </span>
                      <time
                        class="date"
                        datetime="2024-02-20"
                      >
                        20 February 2024
                      </time>
                    </div>
                  </footer>
                </div>
              </article>
            </li>
            <li
              class="newsItem"
            >
              <article
                class="newsCard "
              >
                <a
                  aria-hidden="true"
                  class="imageContainer"
                  href="/i/am/a-very-full-slug"
                  style="background-image: url(http://local-host-test-nice-org.com/f/243782/420x234/7a576133e7/register-for-board-meeting.PNG/m/867x0/filters:format%28jpeg%29:quality%2880%29/m/868x0/filters:quality%2880%29);"
                  tabindex="-1"
                >
                   
                </a>
                <div>
                  <h3
                    class="heading"
                  >
                    <a
                      href="/i/am/a-very-full-slug"
                    >
                      Test news article title
                    </a>
                  </h3>
                  <p>
                    Test intro text
                  </p>
                  <footer>
                    <div>
                      <span
                        class="tag tag--outline"
                        data-component="tag"
                      >
                        News
                      </span>
                      <time
                        class="date"
                        datetime="2024-01-31T15:16:15.19"
                      >
                        31 January 2024
                      </time>
                    </div>
                  </footer>
                </div>
              </article>
            </li>
          </ol>
          <p>
            <a
              href="/news"
            >
              View all news, blogs and podcasts
            </a>
          </p>
        </section>
        <div
          class="links"
          data-tracking="homepage-links"
        >
          <div
            class="linksContainer"
          >
            <div
              class="storyblokRichTextWrapper "
              data-testid="storyblok-rich-text"
            >
              <h2
                class="sbRichtextHeading2"
              >
                A mock h2 heading in a rich text field
              </h2>
              <h3
                class="sbRichtextHeading3"
              >
                A mock h3 heading
              </h3>
              <h4
                class="sbRichtextHeading4"
              >
                A mock h4 heading
              </h4>
              <h5
                class="sbRichtextHeading5"
              >
                A mock h5 heading
              </h5>
              <h6
                class="sbRichtextHeading6"
              >
                A mock h6 heading
              </h6>
              <p>
                Example text content.
              </p>
              <picture>
                <source
                  srcset="http://local-host-test-nice-org.com/f/243782/420x234/7a576133e7/register-for-board-meeting.PNG/m/867x0/filters:format%28jpeg%29:quality%2880%29/m/420x0/filters:format%28avif%29:quality%2880%29"
                  type="image/avif"
                />
                <source
                  srcset="http://local-host-test-nice-org.com/f/243782/420x234/7a576133e7/register-for-board-meeting.PNG/m/867x0/filters:format%28jpeg%29:quality%2880%29/m/420x0/filters:format%28webp%29:quality%2880%29"
                  type="image/webp"
                />
                <img
                  alt="Kitten wearing mittens"
                  data-nimg="1"
                  decoding="async"
                  height="0"
                  loading="lazy"
                  src="/_next/image?url=http%3A%2F%2Flocal-host-test-nice-org.com%2Ff%2F243782%2F420x234%2F7a576133e7%2Fregister-for-board-meeting.PNG%2Fm%2F867x0%2Ffilters%3Aformat%2528jpeg%2529%3Aquality%252880%2529%2Fm%2F420x0%2Ffilters%3Aformat%2528jpeg%2529%3Aquality%252880%2529&w=1080&q=75"
                  srcset="/_next/image?url=http%3A%2F%2Flocal-host-test-nice-org.com%2Ff%2F243782%2F420x234%2F7a576133e7%2Fregister-for-board-meeting.PNG%2Fm%2F867x0%2Ffilters%3Aformat%2528jpeg%2529%3Aquality%252880%2529%2Fm%2F420x0%2Ffilters%3Aformat%2528jpeg%2529%3Aquality%252880%2529&w=640&q=75 1x, /_next/image?url=http%3A%2F%2Flocal-host-test-nice-org.com%2Ff%2F243782%2F420x234%2F7a576133e7%2Fregister-for-board-meeting.PNG%2Fm%2F867x0%2Ffilters%3Aformat%2528jpeg%2529%3Aquality%252880%2529%2Fm%2F420x0%2Ffilters%3Aformat%2528jpeg%2529%3Aquality%252880%2529&w=1080&q=75 2x"
                  style="color: transparent;"
                  width="420"
                />
              </picture>
              <figure
                class="quote"
              >
                <blockquote>
                  <p>
                    A inline quote in the rich text field of a news article. 
                  </p>
                </blockquote>
              </figure>
            </div>
          </div>
        </div>
      </div>
    </main>
    <noscript />
  </div>
</div>
`;
